File: Sources/HealthReminderCore/ReminderEngine.swift
import Foundation

public enum ReminderEngineState: Equatable {
    case tracking
    case pausedByIdle
}

public struct ReminderDefinition: Equatable {
    public let id: String
    public let title: String
    public let body: String
    public let interval: TimeInterval

    public init(
        id: String,
        title: String,
        body: String,
        interval: TimeInterval
    ) {
        self.id = id
        self.title = title
        self.body = body
        self.interval = interval
    }
}

public struct ReminderProgress: Equatable {
    public let definition: ReminderDefinition
    public let elapsedActiveTime: TimeInterval
    public let state: ReminderEngineState

    public var remainingActiveTime: TimeInterval {
        max(0, definition.interval - elapsedActiveTime)
    }
}

public struct ReminderTickResult: Equatable {
    public let remindersToSend: [ReminderDefinition]

    public var shouldSendReminder: Bool {
        !remindersToSend.isEmpty
    }
}

public struct ReminderEngine {
    public let idleThreshold: TimeInterval
    public let tickInterval: TimeInterval

    public private(set) var state: ReminderEngineState
    private var records: [ReminderRecord]

    public init(
        reminders: [ReminderDefinition],
        idleThreshold: TimeInterval,
        tickInterval: TimeInterval
    ) {
        precondition(Set(reminders.map(\.id)).count == reminders.count, "Reminder IDs must be unique.")

        self.idleThreshold = idleThreshold
        self.tickInterval = tickInterval
        self.state = .tracking
        self.records = reminders.map { ReminderRecord(definition: $0) }
    }

    public var reminders: [ReminderProgress] {
        records.map(\.progress)
    }

    public var nextReminder: ReminderProgress? {
        reminders
            .filter { $0.state == .tracking }
            .min { $0.remainingActiveTime < $1.remainingActiveTime }
    }

    public func progress(for reminderID: String) -> ReminderProgress? {
        records.first { $0.definition.id == reminderID }?.progress
    }

    public mutating func tick(idleSeconds: TimeInterval) -> ReminderTickResult {
        if idleSeconds >= idleThreshold {
            updateOverallState(isIdle: true)
            return ReminderTickResult(remindersToSend: [])
        }

        var remindersToSend: [ReminderDefinition] = []

        for index in records.indices {
            if let reminder = records[index].tick(tickInterval: tickInterval) {
                remindersToSend.append(reminder)
            }
        }

        updateOverallState(isIdle: false)
        return ReminderTickResult(remindersToSend: remindersToSend)
    }

    private mutating func updateOverallState(isIdle: Bool) {
        state = isIdle ? .pausedByIdle : .tracking
    }
}

private struct ReminderRecord {
    let definition: ReminderDefinition
    private(set) var elapsedActiveTime: TimeInterval
    private(set) var state: ReminderEngineState

    init(definition: ReminderDefinition) {
        self.definition = definition
        self.elapsedActiveTime = 0
        self.state = .tracking
    }

    var progress: ReminderProgress {
        ReminderProgress(
            definition: definition,
            elapsedActiveTime: elapsedActiveTime,
            state: state
        )
    }

    mutating func tick(tickInterval: TimeInterval) -> ReminderDefinition? {
        state = .tracking
        elapsedActiveTime += tickInterval

        if elapsedActiveTime >= definition.interval {
            reset()
            return definition
        }

        return nil
    }

    mutating func reset() {
        elapsedActiveTime = 0
        state = .tracking
    }
}

File: Windows/HealthReminder.Windows/Core/ReminderEngine.cs
namespace HealthReminder.WindowsApp.Core;

public enum ReminderEngineState
{
    Tracking,
    PausedByIdle
}

public sealed record ReminderDefinition(
    string Id,
    string Title,
    string Body,
    TimeSpan Interval
);

public sealed record ReminderProgress(
    ReminderDefinition Definition,
    TimeSpan ElapsedActiveTime,
    ReminderEngineState State
)
{
    public TimeSpan RemainingActiveTime =>
        Definition.Interval - ElapsedActiveTime > TimeSpan.Zero
            ? Definition.Interval - ElapsedActiveTime
            : TimeSpan.Zero;
}

public sealed record ReminderTickResult(IReadOnlyList<ReminderDefinition> RemindersToSend)
{
    public bool ShouldSendReminder => RemindersToSend.Count > 0;
}

public sealed class ReminderEngine
{
    private readonly List<ReminderRecord> _records;

    public ReminderEngine(
        IReadOnlyList<ReminderDefinition> reminders,
        TimeSpan idleThreshold,
        TimeSpan tickInterval
    )
    {
        if (reminders.Select(reminder => reminder.Id).Distinct().Count() != reminders.Count)
        {
            throw new ArgumentException("Reminder IDs must be unique.", nameof(reminders));
        }

        IdleThreshold = idleThreshold;
        TickInterval = tickInterval;
        State = ReminderEngineState.Tracking;
        _records = reminders.Select(reminder => new ReminderRecord(reminder)).ToList();
    }

    public TimeSpan IdleThreshold { get; }
    public TimeSpan TickInterval { get; }
    public ReminderEngineState State { get; private set; }

    public IReadOnlyList<ReminderProgress> Reminders =>
        _records.Select(record => record.Progress).ToList();

    public ReminderProgress? NextReminder =>
        Reminders
            .Where(progress => progress.State == ReminderEngineState.Tracking)
            .MinBy(progress => progress.RemainingActiveTime);

    public ReminderProgress? ProgressFor(string reminderId) =>
        _records.FirstOrDefault(record => record.Definition.Id == reminderId)?.Progress;

    public ReminderTickResult Tick(TimeSpan idleTime)
    {
        if (idleTime >= IdleThreshold)
        {
            State = ReminderEngineState.PausedByIdle;
            return new ReminderTickResult([]);
        }

        List<ReminderDefinition> remindersToSend = [];

        foreach (ReminderRecord record in _records)
        {
            ReminderDefinition? reminder = record.Tick(TickInterval);
            if (reminder is not null)
            {
                remindersToSend.Add(reminder);
            }
        }

        State = ReminderEngineState.Tracking;
        return new ReminderTickResult(remindersToSend);
    }

    private sealed class ReminderRecord
    {
        public ReminderRecord(ReminderDefinition definition)
        {
            Definition = definition;
            ElapsedActiveTime = TimeSpan.Zero;
            State = ReminderEngineState.Tracking;
        }

        public ReminderDefinition Definition { get; }
        public TimeSpan ElapsedActiveTime { get; private set; }
        public ReminderEngineState State { get; private set; }

        public ReminderProgress Progress =>
            new(Definition, ElapsedActiveTime, State);

        public ReminderDefinition? Tick(TimeSpan tickInterval)
        {
            State = ReminderEngineState.Tracking;
            ElapsedActiveTime += tickInterval;

            if (ElapsedActiveTime < Definition.Interval)
            {
                return null;
            }

            Reset();
            return Definition;
        }

        private void Reset()
        {
            ElapsedActiveTime = TimeSpan.Zero;
            State = ReminderEngineState.Tracking;
        }
    }
}
